File: 977-squares-of-sorted-array/go/main.go
package main

const UintMax = ^uint(0)
const IntMax = int(UintMax >> 1)

func sortedSquares(A []int) []int {
	positive := 0
	res := []int{}

	for positive < len(A) && A[positive] < 0 {
		positive++
	}

	negative := positive - 1

	for positive < len(A) || negative >= 0 {
		a := IntMax
		b := IntMax

		if positive < len(A) {
			a = A[positive] * A[positive]
		}

		if negative >= 0 {
			b = A[negative] * A[negative]
		}

		if a < b {
			res = append(res, a)
			positive++
		} else {
			res = append(res, b)
			negative--
		}
	}

	return res
}

File: 977-squares-of-sorted-array/cpp/main.cpp
class Solution {
 public:
  vector<int> sortedSquares(vector<int>& A) {
    int n = A.size();
    vector<int> sorted(n);
    int positive = 0;
    int negative = 0;

    while (positive < n && A[positive] < 0) {
      ++positive;
    }

    negative = positive - 1;
    int i = 0;
    while (positive < n || negative >= 0) {
      int a = positive < n ? A[positive] * A[positive] : INT_MAX;
      int b = negative >= 0 ? A[negative] * A[negative] : INT_MAX;

      if (a < b) {
        sorted[i++] = a;
        ++positive;
      } else {
        sorted[i++] = b;
        --negative;
      }
    }

    return sorted;
  }
};